hasktorch/src/Torch/Data/StreamedPipeline.hs

```diff
@@ -0,0 +1,216 @@
+  where
+    unwrap output = runEffect $ enumerate prod >-> toOutput' output
+
+streamBatches fold inputBox = foldFromProducer inputs fold
+  where
+    inputs = fromInput' inputBox
+
+foldOverWith' ::
+  forall m dataset seed batch b.
+  (Datastream m seed dataset batch, MonadBaseControl IO m) =>
+  dataset ->
+  ListT m seed ->
+  FoldM m batch b ->
+  m b
+foldOverWith' = foldOverWith defaultDataloaderOpts
+
+foldOverWith ::
+  forall m dataset seed batch b.
+  (Datastream m seed dataset batch, MonadBaseControl IO m) =>
+  DataloaderOpts ->
+  dataset ->
+  ListT m seed ->
+  FoldM m batch b ->
+  m b
+foldOverWith opts@DataloaderOpts {..} dataset seeds fold = do
+  join $ fmap (flip foldFromProducer fold . enumerate) $ makeListT opts dataset seeds
+
+makeListT ::
+  forall batch m dataset seed b.
+  (Datastream m seed dataset batch, MonadBaseControl IO m, MonadBase IO m) =>
+  DataloaderOpts ->
+  dataset ->
+  ListT m seed ->
+  m (ListT m batch)
+makeListT DataloaderOpts {..} dataset seeds = do
+  fmap Select $
+    withOne
+      (bounded numWorkers)
+      (\batchOutput -> readBatches dataset seeds batchOutput)
+      (\input -> pure $ fromInput' input)
+
+makeListT' ::
+  forall batch m f dataset seed b.
+  (Datastream m seed dataset batch, MonadBaseControl IO m, MonadBase IO m, Foldable f) =>
+  DataloaderOpts ->
+  dataset ->
+  f seed ->
+  m (ListT m batch)
+makeListT' DataloaderOpts {..} dataset seeds = do
+  fmap Select $
+    withOne
+      (bounded numWorkers)
+      (\batchOutput -> readBatches' dataset seeds batchOutput)
+      (\input -> (pure $ fromInput' input))
+
+defaultDataloaderOpts = DataloaderOpts {numWorkers = 1}
+
+foldFromProducer :: Monad m => Producer batch m () -> L.FoldM m batch b -> m b
+foldFromProducer prod fold = (L.impurely P.foldM) fold prod
+
+liftedBracket :: MonadBaseControl IO m => m a -> (a -> m b) -> (a -> m c) -> m c
+liftedBracket acquire release action = control $ \runInIO ->
+  bracket
+    (runInIO acquire)
+    (\saved -> runInIO (restoreM saved >>= release))
+    (\saved -> runInIO (restoreM saved >>= action))
+
+withBufferLifted ::
+  (MonadBaseControl IO m) =>
+  Buffer a ->
+  (Output a -> m l) ->
+  (Input a -> m r) ->
+  m (l, r)
+withBufferLifted buffer fOutput fInput =
+  liftedBracket
+    (liftBase $ spawn' buffer)
+    (\(_, _, seal) -> liftBase $ atomically seal)
+    ( \(output, input, seal) ->
+        concurrently
+          (fOutput output `liftedFinally` (liftBase $ atomically seal))
+          (fInput input `liftedFinally` (liftBase $ atomically seal))
+    )
+
+-- We need to cleanup exception handling here more, but withBuffer is too restrictive
+-- for how we use input and output as it ends up sealing too quickly
+withOne ::
+  (MonadBaseControl IO m) =>
+  Buffer a ->
+  (Output a -> m l) ->
+  (Input a -> m r) ->
+  m r
+withOne buffer fOutput fInput =
+  liftedBracket
+    (liftBase $ spawn' buffer)
+    ( \(_, _, seal) -> do liftBase $ print "here"
```

Comment:
In development?

Reply:
https://hackage.haskell.org/package/pipes-concurrency-2.0.12/docs/src/Pipes.Concurrent.html#spawn%27
seal implementation is here.
I will deepen my understanding later.

Reply:
This should be fixed in the current version.